src/server.tsx

```diff
@@ -66,26 +66,37 @@ const buildEpic = async (metadata: EpicMetadata): Promise<ResponseType> => {
     return { html, css };
 };
 
-app.get('/epic', async (req: express.Request, res: express.Response) => {
-    try {
-        const { html, css } = await buildEpic(testData.metadata);
-        const htmlContent = renderHtmlDocument({ html, css });
-        res.send(htmlContent);
-    } catch (error) {
-        console.log('Something went wrong: ', error.message);
-        res.status(500).send({ error: error.message });
-    }
-});
+app.get(
+    '/epic',
+    async (req: express.Request, res: express.Response, next: express.NextFunction) => {
+        try {
+            const { html, css } = await buildEpic(testData.metadata);
+            const htmlContent = renderHtmlDocument({ html, css });
+            res.send(htmlContent);
+        } catch (error) {
+            next(error);
+        }
+    },
+);
+
+app.post(
+    '/epic',
+    async (req: express.Request, res: express.Response, next: express.NextFunction) => {
+        try {
+            const metadata = buildMetadata(req);
+            const { html, css } = await buildEpic(metadata);
+            res.send({ html, css });
+        } catch (error) {
+            next(error);
+        }
+    },
+);
 
-app.post('/epic', async (req: express.Request, res: express.Response) => {
-    try {
-        const metadata = buildMetadata(req);
-        const { html, css } = await buildEpic(metadata);
-        res.send({ html, css });
-    } catch (error) {
-        console.log('Something went wrong: ', error.message);
-        res.status(500).send({ error: error.message });
-    }
+// Error handling middleware in Express needs to take 4 arguments in the handler
+// for it to run when `next()` function is called in the route handler
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+app.use((error: Error, req: express.Request, res: express.Response, next: express.NextFunction) => {
```

Comment:
I guess the position of this code which adds the error handling middleware is important (i.e. it has to be below the get/post handlers above?) I wonder if there's a good way of making that clear/harder to break things if someone moved methods around down the line?

I also wonder if at this point we'd get some value out of having a few integration tests, given that the wiring is getting more complex. I think it might give us additional confidence that everything is working correctly.

Reply:
> Nice! This is looking good. Conceptually I'm slightly torn about whether middleware is the best choice for returning errors consistently. It's elegant for sure, but I also wonder if it's a bit less simple/obvious than having a function that we call directly to return an error response, for example. What do you think?

Fair enough. I'm still not totally convinced either. Doing this in the middleware as we have it adds the benefit of having a single place where error responses are handled and sent to the client. But it does bring the inconvenience of making things slightly less obvious (calling the `next()` function in the route) and having to keep the middleware below all the routes.